src/transformStateWithClones.js

```diff
@@ -1,13 +1,39 @@
-'use strict';
+const ACTION_TYPE_CLEAR = 'clear';
+const ACTION_TYPE_ADD_PROPERTIES = 'addProperties';
+const ACTION_TYPE_REMOVE_PROPERTIES = 'removeProperties';
 
 /**
- * @param {Object} state
- * @param {Object[]} actions
+ * Оновлює стан об'єкта на основі масиву дій і повертає історію змін.
  *
- * @return {Object[]}
+ * @param {Object} initialState - Початковий стан об'єкта.
+ * @param {Object[]} actions - Масив дій, які будуть застосовані до стану.
+ * @returns {Object[]} - Масив станів після кожної дії.
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(initialState, actions) {
+  const stateHistory = [];
+  let currentState = { ...initialState };
+
+  actions.forEach((action) => {
+    if (action.type === ACTION_TYPE_CLEAR) {
+      currentState = {};
+    }
+
+    if (action.type === ACTION_TYPE_ADD_PROPERTIES) {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === ACTION_TYPE_REMOVE_PROPERTIES) {
+      currentState = { ...currentState };
+
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    }
```

Comment:
You should perform the action of pushing the `currentState` to the `stateHistory` array just once after the current action is processed, outside of the `if` or `switch` block. This avoids repeating the same action in all cases and adheres to the [DONT REPEAT YOURSELF] principle.